### ISSUE TITLE
Add support for GNOME 40

### ISSUE THREAD

**Issue by frandieguez** [martin31821/cpupower#145]
**Extension Information**

- Shell Version: GNOME 40
- Extension Version: 23
- Install method: extension.gnome.org
- OS: Fedora 33 with GNOME 40 alfa installed

```sudo dnf copr enable haeckerfelix/gnome-shell-40 && sudo dnf update```

**Describe the bug**

GNOME 40 is already in alpha and it forces the usage of GTK4 for extensions dialogs, please consider adding support for it.

The extensions seems to work so far but when you try to open the preferences dialog of your awesome extension it crashes with the message below 👇

```
Gtk.BuilderError: /home/fran/.local/share/gnome-shell/extensions/cpupower@mko-sl.de/data/cpupower-preferences.glade:88:1 Invalid object type 'GtkButtonBox'

Stack trace:
  CPUPowerPreferences@/home/fran/.local/share/gnome-shell/extensions/cpupower@mko-sl.de/src/preferences.js:53:22
  buildPrefsWidget@/home/fran/.local/share/gnome-shell/extensions/cpupower@mko-sl.de/prefs.js:41:23
  _init@resource:///org/gnome/Shell/Extensions/js/extensionsService.js:209:40
  OpenExtensionPrefsAsync/<@resource:///org/gnome/Shell/Extensions/js/extensionsService.js:122:28
  asyncCallback@resource:///org/gnome/gjs/modules/core/overrides/Gio.js:132:22
  run@resource:///org/gnome/Shell/Extensions/js/dbusService.js:177:20
  main@resource:///org/gnome/Shell/Extensions/js/main.js:19:13
  run@resource:///org/gnome/gjs/modules/script/package.js:222:19
  start@resource:///org/gnome/gjs/modules/script/package.js:206:8
  @/usr/share/gnome-shell/org.gnome.Shell.Extensions:1:17
  
```

Please update you extension to keep enjoying using it. Thanks for your incredible work!

**Comment by martin31821** [comment]
Gnome 4 will most likely be supported after it hits the large distributions. At the moment I'm not able to update yet.

**Comment by fin-ger** [comment]
I wont have time to work on this until April. Also, is it really necessary to temper with my own system to develop for a new Gnome platform? I remember that I was able to run another version of the Gnome Shell within a window in my production setup some time ago. I just can't remember how I managed to achieve this...

**Comment by frandieguez** [comment]
You can test it from GNOME Boxes by selecting the GNOME OS Nightly image

**Comment by fin-ger** [comment]
Hmm, intel p-state will not be available, but I guess this can be mocked with some hacks. As I'm on AMD Ryzen now, I might integrate support for zenstates or ryzenadj along with Gnome 40 support in April (I still have my old intel machine for testing). I might also look into the `/sys/devices/system/cpu/cpufreq/` kernel-API, maybe it's a more general approach which might serve this extension well. I've not decided, yet.

**Comment by mlaverdiere** [comment]
For reference, in case it might be useful, here's the guide to port extensions to GNOME 40: https://gjs.guide/extensions/upgrading/gnome-shell-40.html#metadata-json

Thanks for this extension.

**Comment by mastercaution** [comment]
Gnome 40 landed on Arch yesterday. This issue still exists. 
We need to port our GTK3 glade file to GTK4. The `gtk4-builder-tool` (Reference: https://developer.gnome.org/gtk4/stable/gtk4-builder-tool.html) might come in handy as well as the guide suggested by @mlaverdiere.

The output of `gtk4-builder-tool simplify --3to4 data/cpupower-preferences.glade` looks like a bit of work, and unfortunately I'm pretty busy at the moment.

To get everything else than the preferences window working we just have to add the new Gnome version to the `metadata.json` like described [here](https://gjs.guide/extensions/upgrading/gnome-shell-40.html#metadata-json).

**Comment by fin-ger** [comment]
I'll look into this as soon as I can, currently still busy with my master defense.

I personally think that Gtk4 was pulled into Gnome Extensions too early, as libadwaita is not available yet. I'm less motivated to rebuild the extensions to Gtk4 just to rebuild it again for Gnome 42... So probably the Gnome40 release of this extension will provide less stability and comfort than other releases of cpupower. I'm not quite sure how to handle this properly, I'm also not sure if Gtk4 would run on Ubuntu 18.04/etc. I would love to still support those old Gnome releases with the latest version of this extension as people are still using it. At least Gnome 3.36 should be supported by the latest release. However, I don't know if we can target Gtk4 on these platforms... Maybe there is a way to let Gnome40 run Gtk3 until Gtk3 slowly disappears from distros and Gtk4 can be assumed to be available on older platforms. Does anyone know more about this?

**Comment by pszafer** [comment]
@fin-ger if you target only new release in manifest.json then it should install older version of extension in older Gnome and the newest version in Gnome 40

**Comment by fin-ger** [comment]
Okay, but this would remove support for all other Gnome versions in the latest release. I'm myself still on 3.36 as 3.38 is too unstable to get my daily work done and 40 is not available yet. I would prefer not to ditch all previous Gnome versions.

**Comment by keeferrourke** [comment]
GNOME 40 is officially available through the Fedora 34 release that was made widely available today.

If you need volunteers to get this extension ported, I am dearly missing it in my newly upgraded shell ;)

**Comment by bcclsn** [comment]
so, then, is there any news?

**Comment by fin-ger** [comment]
Working on it, but may take some time to finish. Maybe next week or sth.

**Comment by fin-ger** [comment]
Some news on this: the Gnome OS installation has a broken polkit setup where pkexec isn't able to prompt for user authentication. Therefore, the complete extension is unable to operate. I am now looking for an alternative to develop and try out the Gnome 40 support. Apart from that, Gnome OS seems like a bad choice as I wasn't able to get a dev environment set up (git, shellcheck, make, etc.), leading to continuous copying of new code from the host system paired with the everlasting log-out/log-in procedure to reload the extension code... 

The dev experience of Gnome gets worse and worse with every new release....

**Comment by shroudedcode** [comment]
> ...paired with the everlasting log-out/log-in procedure to reload the extension code...

Is there a reason you can't use <kbd>Ctrl</kbd>+<kbd>F2</kbd> + `r` + <kbd>Enter</kbd>? 🤔

**Comment by fin-ger** [comment]
> > > Is there a reason you can't use Ctrl+F2 + `r` + Enter?
> > 
> > 
> > Because it is not available on Wayland
> 
> Ah, that's a bummer, but maybe not even an issue in this particular case since @fin-ger is already using a virtual machine for development and could [disable Wayland](https://wiki.archlinux.org/title/GDM#Use_Xorg_backend) there. This shouldn't cause any issues as the extension doesn't contain any Wayland/X11-specific code (correct me if I'm wrong).

I already tried that, in Gnome OS selecting `GNOME on Xorg` crashes the session and throws you back to the login screen (running in Gnome Boxes). But, I'm currently running Fedora 34 in a VM, let's see how this goes...

**Comment by fin-ger** [comment]
Well... I invested 2 days in getting a GTK4 preferences window running (I barely managed to get it open). This is mostly due to missing resources on the web (as GTK4 is quite new) and missing glade support which forces me to write ui files by hand. I got tired of it and left it aside, also because most changes were breaking for my current 3.36 setup.

Just right now I hacked together a tiny wrapper for the `prefs.js` file which is causing the whole problem of forcing me to use GTK4. I polyfilled the `imports.misc` calls and managed to get the GTK3 preferences dialog running and fully functional on Fedora 34 with Gnome 40 in ~45 lines of js. This solution feels really hacky, but it works... so might be worth some testing I guess?

I will assemble a fully working version tomorrow and commit it to git, right now it's a bunch of randomly stitched scripts in `lg` and an interactive `gjs` shell xD

**Comment by fin-ger** [comment]
Successfully rebased Gnome 40 support on master. Now includes changes of #142. Will prepare a PR now

**Comment by fin-ger** [comment]
Part of #151

**Comment by stefangweichinger** [comment]
I installed the current version via cloning and "make install ...", in a Fedora 34 / GNOME 40 setup.

The addon works, I can use the slider from #142, but clicking on "Settings" fails with:

```
The settings of extension cpupower@mko-sl.de had an error:

Gtk.BuilderError: /home/sgw/.local/share/gnome-shell/extensions/cpupower@mko-sl.de/data/cpupower-preferences.glade:85:1 Invalid object type 'GtkButtonBox'
```

Stack trace:
```
CPUPowerPreferences@/home/sgw/.local/share/gnome-shell/extensions/cpupower@mko-sl.de/src/preferences.js:54:22
buildPrefsWidget@/home/sgw/.local/share/gnome-shell/extensions/cpupower@mko-sl.de/prefs.js:41:23
_init@resource:///org/gnome/Shell/Extensions/js/extensionsService.js:209:40
OpenExtensionPrefsAsync/<@resource:///org/gnome/Shell/Extensions/js/extensionsService.js:122:28
asyncCallback@resource:///org/gnome/gjs/modules/core/overrides/Gio.js:115:22
run@resource:///org/gnome/Shell/Extensions/js/dbusService.js:177:20
main@resource:///org/gnome/Shell/Extensions/js/main.js:19:13
run@resource:///org/gnome/gjs/modules/script/package.js:206:19
start@resource:///org/gnome/gjs/modules/script/package.js:190:8
@/usr/share/gnome-shell/org.gnome.Shell.Extensions:1:17
```

I thought I should report that back to you. Thanks.

**Comment by fin-ger** [comment]
@stefangweichinger It looks like you're not on the `feature/modular-cpufreqctl-backends` branch. The master branch has no support for Gnome 40 yet, the PR #151 is still pending.

**Comment by fin-ger** [comment]
We're currently looking for reviews for #151 regarding the source code and compatibility with the major distros. There have been found some major issues in Gnome <3.34 installs already, but I haven't had the time to fix these issues yet. Feel free to look at the code and give feedback :D

**Comment by stefangweichinger** [comment]
> @stefangweichinger It looks like you're not on the `feature/modular-cpufreqctl-backends` branch. The master branch has no support for Gnome 40 yet, the PR #151 is still pending.

Ah, right, I installed from that branch now, and settings work now. Great, thanks. Sorry for the noise ;-)

**Comment by fin-ger** [comment]
@stefangweichinger Please note that when installing from e.g. feature branches you are required to uninstall the tool and polkit rules and reinstall it to make the extension work flawlessly. The update procedure for polkit rules and the `cpufreqctl` tool is based on the version number which remains unchanged in non-releases and therefore does not trigger the update procedure.

**Comment by stefangweichinger** [comment]
@fin-ger Oh, I see, thanks. I will maybe re-install later. For now things seem to work .. and I am lazy/busy right now ;-)

**Comment by bcclsn** [comment]
I've installed this release, it seems to work fine except the icon and the current frequency in the top bar, which are not displayed... as you can see in the attached image.
[image](https://postimg.cc/CB5DYJNm)
furthermore, a generic profile is loaded at startup instead of the one set for battery use. auto switch is enabled.

I'm on Archlinux with gnome 40.1

**Comment by fin-ger** [comment]
@bcclsn The "Current Frequency" seems to be locked at 800 MHz, so something is absolutely not running smoothly here. This generally means that communication with the backend fails at a very early stage in initialization. Have you uninstalled and installed the tool and polkit rule after switching to the branch?

**Comment by bcclsn** [comment]
mmm yes, you ar right...
i don't know if it is relevant from or not, but from cli the frequency changes correctly.

yes, I followed the whole procedure

**Comment by fin-ger** [comment]
The cli received a major update, so it is important you have the new version installed otherwise it won't work. The development version isn't able to check for compatibility of the cli and the extension code as this is solely based on the version number.

Can you provide the the output of `journalctl` relevant to cpupower/gnome-shell?

**Comment by bcclsn** [comment]
seems there is an error with metadata
`
mag 23 15:35:37 hostname gnome-shell[765]: JS ERROR: Could not load extension cpupower@mko-sl.de-backup: Error: Missing metadata.json`

**Comment by bcclsn** [comment]
only cpupower
`mag 23 15:56:58 dreamspark pkexec[12908]: bcclsn: Executing command [USER=root] [TTY=unknown] [CWD=/home/bcclsn/.local/share/gnome-shell/extensions/cpupower@mko-sl.de] [COMMAND=/usr/local/bin/cpufreqctl-bcclsn --backend intel_pstate --format json min set 18]
                                               powerActions@/home/bcclsn/.local/share/gnome-shell/extensions/cpupower@mko-sl.de/src/indicator.js:136:33
                                               enable@/home/bcclsn/.local/share/gnome-shell/extensions/cpupower@mko-sl.de/src/indicator.js:112:14
                                               enableIndicator@/home/bcclsn/.local/share/gnome-shell/extensions/cpupower@mko-sl.de/extension.js:48:23
                                               enable/<@/home/bcclsn/.local/share/gnome-shell/extensions/cpupower@mko-sl.de/extension.js:88:13
                                               spawnProcessCheckExitCode/<@/home/bcclsn/.local/share/gnome-shell/extensions/cpupower@mko-sl.de/src/utils.js:63:21
mag 23 15:56:58 dreamspark pkexec[12910]: bcclsn: Executing command [USER=root] [TTY=unknown] [CWD=/home/bcclsn/.local/share/gnome-shell/extensions/cpupower@mko-sl.de] [COMMAND=/usr/local/bin/cpufreqctl-bcclsn --backend intel_pstate --format json turbo set off]
mag 23 15:56:58 dreamspark pkexec[12909]: bcclsn: Executing command [USER=root] [TTY=unknown] [CWD=/home/bcclsn/.local/share/gnome-shell/extensions/cpupower@mko-sl.de] [COMMAND=/usr/local/bin/cpufreqctl-bcclsn --backend intel_pstate --format json max set 18]
mag 23 15:56:58 dreamspark pkexec[12913]: bcclsn: Executing command [USER=root] [TTY=unknown] [CWD=/home/bcclsn/.local/share/gnome-shell/extensions/cpupower@mko-sl.de] [COMMAND=/usr/local/bin/cpufreqctl-bcclsn --backend intel_pstate --format json info frequencies]
mag 23 15:57:00 dreamspark pkexec[12948]: bcclsn: Executing command [USER=root] [TTY=unknown] [CWD=/home/bcclsn/.local/share/gnome-shell/extensions/cpupower@mko-sl.de] [COMMAND=/usr/local/bin/cpufreqctl-bcclsn --backend intel_pstate --format json min set 18]
mag 23 15:57:00 dreamspark pkexec[12949]: bcclsn: Executing command [USER=root] [TTY=unknown] [CWD=/home/bcclsn/.local/share/gnome-shell/extensions/cpupower@mko-sl.de] [COMMAND=/usr/local/bin/cpufreqctl-bcclsn --backend intel_pstate --format json max set 18]
mag 23 15:57:06 dreamspark pkexec[12967]: bcclsn: Executing command [USER=root] [TTY=unknown] [CWD=/home/bcclsn/.local/share/gnome-shell/extensions/cpupower@mko-sl.de] [COMMAND=/usr/local/bin/cpufreqctl-bcclsn --backend intel_pstate --format json min set 50]
mag 23 15:57:06 dreamspark pkexec[12968]: bcclsn: Executing command [USER=root] [TTY=unknown] [CWD=/home/bcclsn/.local/share/gnome-shell/extensions/cpupower@mko-sl.de] [COMMAND=/usr/local/bin/cpufreqctl-bcclsn --backend intel_pstate --format json max set 100]
mag 23 15:57:06 dreamspark pkexec[12971]: bcclsn: Executing command [USER=root] [TTY=unknown] [CWD=/home/bcclsn/.local/share/gnome-shell/extensions/cpupower@mko-sl.de] [COMMAND=/usr/local/bin/cpufreqctl-bcclsn --backend intel_pstate --format json turbo set on]
mag 23 15:57:06 dreamspark pkexec[12976]: bcclsn: Executing command [USER=root] [TTY=unknown] [CWD=/home/bcclsn/.local/share/gnome-shell/extensions/cpupower@mko-sl.de] [COMMAND=/usr/local/bin/cpufreqctl-bcclsn --backend intel_pstate --format json min set 50]
mag 23 15:57:06 dreamspark pkexec[12979]: bcclsn: Executing command [USER=root] [TTY=unknown] [CWD=/home/bcclsn/.local/share/gnome-shell/extensions/cpupower@mko-sl.de] [COMMAND=/usr/local/bin/cpufreqctl-bcclsn --backend intel_pstate --format json max set 100]
`
gnome-shell
`mag 23 15:55:13 dreamspark systemd[698]: Started Application launched by gnome-shell.
mag 23 15:55:21 dreamspark gnome-shell[8390]: meta_window_set_stack_position_no_sync: assertion 'window->stack_position >= 0' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: meta_later_add: assertion 'display' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: meta_later_remove: assertion 'display' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: meta_later_remove: assertion 'display' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: JS ERROR: TypeError: adjustment is null
mag 23 15:56:40 dreamspark gnome-shell[8390]: meta_later_remove: assertion 'display' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: JS ERROR: TypeError: adjustment is null
mag 23 15:56:40 dreamspark gnome-shell[8390]: meta_later_remove: assertion 'display' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: JS ERROR: TypeError: adjustment is null
mag 23 15:56:40 dreamspark gnome-shell[8390]: meta_later_remove: assertion 'display' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: JS ERROR: TypeError: adjustment is null
mag 23 15:56:40 dreamspark gnome-shell[8390]: meta_later_remove: assertion 'display' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: JS ERROR: TypeError: adjustment is null
mag 23 15:56:40 dreamspark gnome-shell[8390]: meta_later_remove: assertion 'display' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: JS ERROR: TypeError: adjustment is null
mag 23 15:56:40 dreamspark gnome-shell[8390]: JS ERROR: TypeError: adjustment is null
mag 23 15:56:40 dreamspark gnome-shell[8390]: meta_later_add: assertion 'display' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: meta_later_add: assertion 'display' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: meta_later_add: assertion 'display' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: meta_later_add: assertion 'display' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: meta_later_add: assertion 'display' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: meta_later_add: assertion 'display' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: meta_later_add: assertion 'display' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: meta_later_add: assertion 'display' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: meta_later_add: assertion 'display' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: meta_later_add: assertion 'display' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: meta_later_add: assertion 'display' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: meta_later_add: assertion 'display' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: meta_later_add: assertion 'display' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: meta_later_add: assertion 'display' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: JS ERROR: TypeError: container is null
mag 23 15:56:40 dreamspark gnome-shell[8390]: clutter_actor_iter_next: assertion 'ri->age == ri->root->priv->age' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: instance of invalid non-instantiatable type '(null)'
mag 23 15:56:40 dreamspark gnome-shell[8390]: g_signal_handlers_disconnect_matched: assertion 'G_TYPE_CHECK_INSTANCE (instance)' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: instance of invalid non-instantiatable type '(null)'
mag 23 15:56:40 dreamspark gnome-shell[8390]: g_signal_handler_disconnect: assertion 'G_TYPE_CHECK_INSTANCE (instance)' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: instance of invalid non-instantiatable type '(null)'
mag 23 15:56:40 dreamspark gnome-shell[8390]: g_signal_handler_disconnect: assertion 'G_TYPE_CHECK_INSTANCE (instance)' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: instance of invalid non-instantiatable type '(null)'
mag 23 15:56:40 dreamspark gnome-shell[8390]: g_signal_handler_disconnect: assertion 'G_TYPE_CHECK_INSTANCE (instance)' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: instance of invalid non-instantiatable type '(null)'
mag 23 15:56:40 dreamspark gnome-shell[8390]: g_signal_handler_disconnect: assertion 'G_TYPE_CHECK_INSTANCE (instance)' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: instance of invalid non-instantiatable type '(null)'
mag 23 15:56:40 dreamspark gnome-shell[8390]: g_signal_handler_disconnect: assertion 'G_TYPE_CHECK_INSTANCE (instance)' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: instance of invalid non-instantiatable type '(null)'
mag 23 15:56:40 dreamspark gnome-shell[8390]: g_signal_handler_disconnect: assertion 'G_TYPE_CHECK_INSTANCE (instance)' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: instance of invalid non-instantiatable type '(null)'
mag 23 15:56:40 dreamspark gnome-shell[8390]: g_signal_handler_disconnect: assertion 'G_TYPE_CHECK_INSTANCE (instance)' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: instance of invalid non-instantiatable type '(null)'
mag 23 15:56:40 dreamspark gnome-shell[8390]: g_signal_handler_disconnect: assertion 'G_TYPE_CHECK_INSTANCE (instance)' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: instance of invalid non-instantiatable type '(null)'
mag 23 15:56:40 dreamspark gnome-shell[8390]: g_signal_handler_disconnect: assertion 'G_TYPE_CHECK_INSTANCE (instance)' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: instance of invalid non-instantiatable type '(null)'
mag 23 15:56:40 dreamspark gnome-shell[8390]: g_signal_handler_disconnect: assertion 'G_TYPE_CHECK_INSTANCE (instance)' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: instance of invalid non-instantiatable type '(null)'
mag 23 15:56:40 dreamspark gnome-shell[8390]: g_signal_handler_disconnect: assertion 'G_TYPE_CHECK_INSTANCE (instance)' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: instance of invalid non-instantiatable type '(null)'
mag 23 15:56:40 dreamspark gnome-shell[8390]: g_signal_handler_disconnect: assertion 'G_TYPE_CHECK_INSTANCE (instance)' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: instance of invalid non-instantiatable type '(null)'
mag 23 15:56:40 dreamspark gnome-shell[8390]: == Stack trace for context 0x55fe900c01b0 ==
mag 23 15:56:40 dreamspark gnome-shell[8390]: == Stack trace for context 0x55fe900c01b0 ==
mag 23 15:56:40 dreamspark gnome-shell[8390]: g_signal_handler_disconnect: assertion 'G_TYPE_CHECK_INSTANCE (instance)' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: == Stack trace for context 0x55fe900c01b0 ==
mag 23 15:56:40 dreamspark gnome-shell[8390]: instance of invalid non-instantiatable type '(null)'
mag 23 15:56:40 dreamspark gnome-shell[8390]: g_signal_handler_disconnect: assertion 'G_TYPE_CHECK_INSTANCE (instance)' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: instance of invalid non-instantiatable type '(null)'
mag 23 15:56:40 dreamspark gnome-shell[8390]: g_signal_handler_disconnect: assertion 'G_TYPE_CHECK_INSTANCE (instance)' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: Attempting to call back into JSAPI during the sweeping phase of GC. This is most likely caused by not destroying a Clutter actor or Gtk+ widget with ::destroy signals connected, but can also be caused by using the destroy(), dispose(), or remove() vfuncs. Because it would crash the application, it has been blocked and the JS callback not invoked.
mag 23 15:56:40 dreamspark gnome-shell[8390]: The offending signal was notify on NMDeviceWifi 0x55fe91fb03c0.
mag 23 15:56:40 dreamspark gnome-shell[8390]: Attempting to call back into JSAPI during the sweeping phase of GC. This is most likely caused by not destroying a Clutter actor or Gtk+ widget with ::destroy signals connected, but can also be caused by using the destroy(), dispose(), or remove() vfuncs. Because it would crash the application, it has been blocked and the JS callback not invoked.
mag 23 15:56:40 dreamspark gnome-shell[8390]: The offending signal was notify on NMActiveConnection 0x55fe9093b180.
mag 23 15:56:40 dreamspark gnome-shell[8390]: Attempting to call back into JSAPI during the sweeping phase of GC. This is most likely caused by not destroying a Clutter actor or Gtk+ widget with ::destroy signals connected, but can also be caused by using the destroy(), dispose(), or remove() vfuncs. Because it would crash the application, it has been blocked and the JS callback not invoked.
mag 23 15:56:40 dreamspark gnome-shell[8390]: The offending signal was state-changed on NMDeviceWifi 0x55fe91fb03c0.
mag 23 15:56:40 dreamspark gnome-shell[8390]: instance of invalid non-instantiatable type '(null)'
mag 23 15:56:40 dreamspark gnome-shell[8390]: g_signal_handler_disconnect: assertion 'G_TYPE_CHECK_INSTANCE (instance)' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: instance of invalid non-instantiatable type '(null)'
mag 23 15:56:40 dreamspark gnome-shell[8390]: g_signal_handler_disconnect: assertion 'G_TYPE_CHECK_INSTANCE (instance)' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: instance of invalid non-instantiatable type '(null)'
mag 23 15:56:40 dreamspark gnome-shell[8390]: g_signal_handler_disconnect: assertion 'G_TYPE_CHECK_INSTANCE (instance)' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: instance of invalid non-instantiatable type '(null)'
mag 23 15:56:40 dreamspark gnome-shell[8390]: g_signal_handler_disconnect: assertion 'G_TYPE_CHECK_INSTANCE (instance)' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: instance of invalid non-instantiatable type '(null)'
mag 23 15:56:40 dreamspark gnome-shell[8390]: g_signal_handler_disconnect: assertion 'G_TYPE_CHECK_INSTANCE (instance)' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: instance of invalid non-instantiatable type '(null)'
mag 23 15:56:40 dreamspark gnome-shell[8390]: g_signal_handler_disconnect: assertion 'G_TYPE_CHECK_INSTANCE (instance)' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: instance of invalid non-instantiatable type '(null)'
mag 23 15:56:40 dreamspark gnome-shell[8390]: g_signal_handler_disconnect: assertion 'G_TYPE_CHECK_INSTANCE (instance)' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: instance of invalid non-instantiatable type '(null)'
mag 23 15:56:40 dreamspark gnome-shell[8390]: g_signal_handler_disconnect: assertion 'G_TYPE_CHECK_INSTANCE (instance)' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: instance of invalid non-instantiatable type '(null)'
mag 23 15:56:40 dreamspark gnome-shell[8390]: g_signal_handler_disconnect: assertion 'G_TYPE_CHECK_INSTANCE (instance)' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: instance of invalid non-instantiatable type '(null)'
mag 23 15:56:40 dreamspark gnome-shell[8390]: g_signal_handler_disconnect: assertion 'G_TYPE_CHECK_INSTANCE (instance)' failed
mag 23 15:56:40 dreamspark gnome-shell[8390]: Attempting to call back into JSAPI during the sweeping phase of GC. This is most likely caused by not destroying a Clutter actor or Gtk+ widget with ::destroy signals connected, but can also be caused by using the destroy(), dispose(), or remove() vfuncs. Because it would crash the application, it has been blocked and the JS callback not invoked.
mag 23 15:56:40 dreamspark gnome-shell[8390]: The offending signal was destroy on StWidget 0x55fe90bbcb40.
mag 23 15:56:40 dreamspark audit[8390]: ANOM_ABEND auid=1000 uid=1000 gid=1000 ses=4 pid=8390 comm="gnome-shell" exe="/usr/bin/gnome-shell" sig=11 res=1
mag 23 15:56:40 dreamspark gnome-shell[8390]: == Stack trace for context 0x55fe900c01b0 ==
mag 23 15:56:40 dreamspark gnome-shell[8390]: == Stack trace for context 0x55fe900c01b0 ==
mag 23 15:56:40 dreamspark gnome-shell[8390]: == Stack trace for context 0x55fe900c01b0 ==
mag 23 15:56:40 dreamspark gnome-shell[8390]: Attempting to call back into JSAPI during the sweeping phase of GC. This is most likely caused by not destroying a Clutter actor or Gtk+ widget with ::destroy signals connected, but can also be caused by using the destroy(), dispose(), or remove() vfuncs. Because it would crash the application, it has been blocked and the JS callback not invoked.
mag 23 15:56:40 dreamspark gnome-shell[8390]: The offending signal was destroy on StBoxLayout 0x55fe90bbefd0.
mag 23 15:56:40 dreamspark gnome-shell[8390]: Attempting to call back into JSAPI during the sweeping phase of GC. This is most likely caused by not destroying a Clutter actor or Gtk+ widget with ::destroy signals connected, but can also be caused by using the destroy(), dispose(), or remove() vfuncs. Because it would crash the application, it has been blocked and the JS callback not invoked.
mag 23 15:56:40 dreamspark gnome-shell[8390]: The offending signal was destroy on Gjs_ui_panel_Panel 0x55fe91084830.
mag 23 15:56:40 dreamspark kernel: gnome-shell[8390]: segfault at 0 ip 0000000000000000 sp 00007ffd61f3e0d8 error 14
mag 23 15:56:40 dreamspark kernel: audit: type=1701 audit(1621778200.312:302): auid=1000 uid=1000 gid=1000 ses=4 pid=8390 comm="gnome-shell" exe="/usr/bin/gnome-shell" sig=11 res=1
mag 23 15:56:41 dreamspark systemd-coredump[12403]: Process 8390 (gnome-shell) of user 1000 dumped core.
mag 23 15:56:42 dreamspark gnome-shell[12445]: Adding device '/dev/dri/card0' (i915) using atomic mode setting.
mag 23 15:56:42 dreamspark gnome-shell[12445]: Boot VGA GPU /dev/dri/card0 selected as primary
mag 23 15:56:43 dreamspark gnome-shell[12445]: Using public X11 display :1024, (using :1025 for managed services)
mag 23 15:56:43 dreamspark gnome-shell[12445]: Using Wayland display name 'wayland-0'
mag 23 15:56:43 dreamspark /usr/lib/gdm-wayland-session[12436]: dbus-daemon[12436]: [session uid=120 pid=12436] Activating service name='org.a11y.Bus' requested by ':1.4' (uid=120 pid=12445 comm="/usr/bin/gnome-shell ")
mag 23 15:56:43 dreamspark gnome-shell[12445]: Skipping parental controls support as it’s disabled
mag 23 15:56:43 dreamspark gnome-shell[12445]: Unset XDG_SESSION_ID, getCurrentSessionProxy() called outside a user session. Asking logind directly.
mag 23 15:56:43 dreamspark gnome-shell[12445]: Will monitor session 8
mag 23 15:56:43 dreamspark dbus-daemon[341]: [system] Activating via systemd: service name='org.freedesktop.locale1' unit='dbus-org.freedesktop.locale1.service' requested by ':1.220' (uid=120 pid=12445 comm="/usr/bin/gnome-shell ")
mag 23 15:56:43 dreamspark gnome-shell[12445]: Failed to launch ibus-daemon: Esecuzione del processo figlio «ibus-daemon» non riuscita (File o directory non esistente)
mag 23 15:56:43 dreamspark /usr/lib/gdm-wayland-session[12436]: dbus-daemon[12436]: [session uid=120 pid=12436] Activating service name='org.freedesktop.impl.portal.PermissionStore' requested by ':1.3' (uid=120 pid=12445 comm="/usr/bin/gnome-shell ")
mag 23 15:56:43 dreamspark dbus-daemon[341]: [system] Activating via systemd: service name='org.freedesktop.GeoClue2' unit='geoclue.service' requested by ':1.220' (uid=120 pid=12445 comm="/usr/bin/gnome-shell ")
mag 23 15:56:43 dreamspark polkitd[404]: Registered Authentication Agent for unix-session:8 (system bus name :1.220 [/usr/bin/gnome-shell], object path /org/freedesktop/PolicyKit1/AuthenticationAgent, locale it_IT.UTF-8)
mag 23 15:56:44 dreamspark /usr/lib/gdm-wayland-session[12436]: dbus-daemon[12436]: [session uid=120 pid=12436] Activating service name='org.gnome.Shell.Notifications' requested by ':1.3' (uid=120 pid=12445 comm="/usr/bin/gnome-shell ")
mag 23 15:56:44 dreamspark /usr/lib/gdm-wayland-session[12470]: dbus-daemon[12470]: Activating service name='org.a11y.atspi.Registry' requested by ':1.0' (uid=120 pid=12445 comm="/usr/bin/gnome-shell ")
mag 23 15:56:44 dreamspark gnome-shell[12445]: Error looking up permission: GDBus.Error:org.freedesktop.portal.Error.NotFound: No entry for geolocation
mag 23 15:56:44 dreamspark gnome-shell[12445]: Failed to launch ibus-daemon: Esecuzione del processo figlio «ibus-daemon» non riuscita (File o directory non esistente)
mag 23 15:56:44 dreamspark gnome-shell[12445]: Failed to init X11 display: Unknown error
mag 23 15:56:45 dreamspark gnome-shell[12445]: Registering session with GDM
mag 23 15:56:56 dreamspark gnome-shell[12699]: Adding device '/dev/dri/card0' (i915) using atomic mode setting.
mag 23 15:56:56 dreamspark gnome-shell[12699]: Boot VGA GPU /dev/dri/card0 selected as primary
mag 23 15:56:56 dreamspark gnome-shell[12699]: Using public X11 display :0, (using :1 for managed services)
mag 23 15:56:56 dreamspark gnome-shell[12699]: Using Wayland display name 'wayland-0'
mag 23 15:56:56 dreamspark dbus-daemon[12658]: [session uid=1000 pid=12658] Activating via systemd: service name='org.a11y.Bus' unit='at-spi-dbus-bus.service' requested by ':1.12' (uid=1000 pid=12699 comm="/usr/bin/gnome-shell ")
mag 23 15:56:57 dreamspark gnome-shell[12699]: Skipping parental controls support as it’s disabled
mag 23 15:56:57 dreamspark gnome-shell[12699]: Unset XDG_SESSION_ID, getCurrentSessionProxy() called outside a user session. Asking logind directly.
mag 23 15:56:57 dreamspark gnome-shell[12699]: Will monitor session 10
mag 23 15:56:57 dreamspark gnome-shell[12699]: Failed to launch ibus-daemon: Esecuzione del processo figlio «ibus-daemon» non riuscita (File o directory non esistente)
mag 23 15:56:57 dreamspark dbus-daemon[12658]: [session uid=1000 pid=12658] Activating via systemd: service name='org.freedesktop.impl.portal.PermissionStore' unit='xdg-permission-store.service' requested by ':1.11' (uid=1000 pid=12699 comm="/usr/bin/gnome-shell ")
mag 23 15:56:57 dreamspark dbus-daemon[12658]: [session uid=1000 pid=12658] Activating service name='org.gnome.Shell.CalendarServer' requested by ':1.11' (uid=1000 pid=12699 comm="/usr/bin/gnome-shell ")
mag 23 15:56:57 dreamspark org.gnome.Shell.CalendarServer[12737]: /usr/lib/gnome-shell-calendar-server: error while loading shared libraries: libecal-2.0.so.1: cannot open shared object file: No such file or directory
mag 23 15:56:57 dreamspark polkitd[404]: Registered Authentication Agent for unix-session:10 (system bus name :1.246 [/usr/bin/gnome-shell], object path /org/freedesktop/PolicyKit1/AuthenticationAgent, locale it_IT.UTF-8)
mag 23 15:56:57 dreamspark gnome-shell[12699]: Telepathy is not available, chat integration will be disabled.
mag 23 15:56:57 dreamspark dbus-daemon[12658]: [session uid=1000 pid=12658] Activating via systemd: service name='org.gtk.vfs.Daemon' unit='gvfs-daemon.service' requested by ':1.11' (uid=1000 pid=12699 comm="/usr/bin/gnome-shell ")
mag 23 15:56:57 dreamspark dbus-daemon[12658]: [session uid=1000 pid=12658] Activating via systemd: service name='org.gtk.vfs.UDisks2VolumeMonitor' unit='gvfs-udisks2-volume-monitor.service' requested by ':1.11' (uid=1000 pid=12699 comm="/usr/bin/gnome-shell ")
mag 23 15:56:57 dreamspark dbus-daemon[12658]: [session uid=1000 pid=12658] Activating via systemd: service name='org.gtk.vfs.MTPVolumeMonitor' unit='gvfs-mtp-volume-monitor.service' requested by ':1.11' (uid=1000 pid=12699 comm="/usr/bin/gnome-shell ")
mag 23 15:56:57 dreamspark gnome-shell[12699]: Extension user-theme@gnome-shell-extensions.gcampax.github.com already installed in /home/bcclsn/.local/share/gnome-shell/extensions/user-theme@gnome-shell-extensions.gcampax.github.com. /usr/share/gnome-shell/extensions/user-theme@gnome-shell-extensions.gcampax.github.com will not be loaded
mag 23 15:56:57 dreamspark at-spi-bus-launcher[12725]: dbus-daemon[12725]: Activating service name='org.a11y.atspi.Registry' requested by ':1.0' (uid=1000 pid=12699 comm="/usr/bin/gnome-shell ")
mag 23 15:56:57 dreamspark dbus-daemon[12658]: [session uid=1000 pid=12658] Activating service name='org.gnome.Shell.Notifications' requested by ':1.11' (uid=1000 pid=12699 comm="/usr/bin/gnome-shell ")
mag 23 15:56:57 dreamspark gnome-shell[12768]: Installed
mag 23 15:56:57 dreamspark gnome-shell[12699]: Error loading calendars: Errore nel chiamare StartServiceByName per org.gnome.Shell.CalendarServer: Process org.gnome.Shell.CalendarServer exited with status 127
mag 23 15:56:58 dreamspark gnome-shell[12699]: Error looking up permission: GDBus.Error:org.freedesktop.portal.Error.NotFound: No entry for geolocation
mag 23 15:56:58 dreamspark pkexec[12908]: bcclsn: Executing command [USER=root] [TTY=unknown] [CWD=/home/bcclsn/.local/share/gnome-shell/extensions/cpupower@mko-sl.de] [COMMAND=/usr/local/bin/cpufreqctl-bcclsn --backend intel_pstate --format json min set 18]
mag 23 15:56:58 dreamspark gnome-shell[12699]: Power state changed: discharging
mag 23 15:56:58 dreamspark gnome-shell[12699]: JS ERROR: TypeError: this.profiles is undefined
                                               powerActions@/home/bcclsn/.local/share/gnome-shell/extensions/cpupower@mko-sl.de/src/indicator.js:136:33
                                               enable@/home/bcclsn/.local/share/gnome-shell/extensions/cpupower@mko-sl.de/src/indicator.js:112:14
                                               enableIndicator@/home/bcclsn/.local/share/gnome-shell/extensions/cpupower@mko-sl.de/extension.js:48:23
                                               enable/<@/home/bcclsn/.local/share/gnome-shell/extensions/cpupower@mko-sl.de/extension.js:88:13
                                               spawnProcessCheckExitCode/<@/home/bcclsn/.local/share/gnome-shell/extensions/cpupower@mko-sl.de/src/utils.js:63:21
mag 23 15:56:58 dreamspark pkexec[12910]: bcclsn: Executing command [USER=root] [TTY=unknown] [CWD=/home/bcclsn/.local/share/gnome-shell/extensions/cpupower@mko-sl.de] [COMMAND=/usr/local/bin/cpufreqctl-bcclsn --backend intel_pstate --format json turbo set off]
mag 23 15:56:58 dreamspark pkexec[12909]: bcclsn: Executing command [USER=root] [TTY=unknown] [CWD=/home/bcclsn/.local/share/gnome-shell/extensions/cpupower@mko-sl.de] [COMMAND=/usr/local/bin/cpufreqctl-bcclsn --backend intel_pstate --format json max set 18]
mag 23 15:56:58 dreamspark pkexec[12913]: bcclsn: Executing command [USER=root] [TTY=unknown] [CWD=/home/bcclsn/.local/share/gnome-shell/extensions/cpupower@mko-sl.de] [COMMAND=/usr/local/bin/cpufreqctl-bcclsn --backend intel_pstate --format json info frequencies]
mag 23 15:57:00 dreamspark pkexec[12948]: bcclsn: Executing command [USER=root] [TTY=unknown] [CWD=/home/bcclsn/.local/share/gnome-shell/extensions/cpupower@mko-sl.de] [COMMAND=/usr/local/bin/cpufreqctl-bcclsn --backend intel_pstate --format json min set 18]
mag 23 15:57:00 dreamspark pkexec[12949]: bcclsn: Executing command [USER=root] [TTY=unknown] [CWD=/home/bcclsn/.local/share/gnome-shell/extensions/cpupower@mko-sl.de] [COMMAND=/usr/local/bin/cpufreqctl-bcclsn --backend intel_pstate --format json max set 18]
mag 23 15:57:01 dreamspark gnome-shell[12699]: Some code called array.toString() on a Uint8Array instance. Previously this would have interpreted the bytes of the array as a string, but that is nonstandard. In the future this will return the bytes as comma-separated digits. For the time being, the old behavior has been preserved, but please fix your code anyway to explicitly call ByteArray.toString(array).
                                               1 parseStat() ["/home/bcclsn/.local/share/gnome-shell/extensions/netspeedsimplified@prateekmedia.extension/extension.js":301:19]
mag 23 15:57:02 dreamspark gnome-shell[12699]: GNOME Shell started at Sun May 23 2021 15:56:57 GMT+0200 (Ora legale dell’Europa centrale)
mag 23 15:57:02 dreamspark gnome-shell[12699]: Registering session with GDM
mag 23 15:57:06 dreamspark pkexec[12967]: bcclsn: Executing command [USER=root] [TTY=unknown] [CWD=/home/bcclsn/.local/share/gnome-shell/extensions/cpupower@mko-sl.de] [COMMAND=/usr/local/bin/cpufreqctl-bcclsn --backend intel_pstate --format json min set 50]
mag 23 15:57:06 dreamspark pkexec[12968]: bcclsn: Executing command [USER=root] [TTY=unknown] [CWD=/home/bcclsn/.local/share/gnome-shell/extensions/cpupower@mko-sl.de] [COMMAND=/usr/local/bin/cpufreqctl-bcclsn --backend intel_pstate --format json max set 100]
mag 23 15:57:06 dreamspark pkexec[12971]: bcclsn: Executing command [USER=root] [TTY=unknown] [CWD=/home/bcclsn/.local/share/gnome-shell/extensions/cpupower@mko-sl.de] [COMMAND=/usr/local/bin/cpufreqctl-bcclsn --backend intel_pstate --format json turbo set on]
mag 23 15:57:06 dreamspark pkexec[12976]: bcclsn: Executing command [USER=root] [TTY=unknown] [CWD=/home/bcclsn/.local/share/gnome-shell/extensions/cpupower@mko-sl.de] [COMMAND=/usr/local/bin/cpufreqctl-bcclsn --backend intel_pstate --format json min set 50]
mag 23 15:57:06 dreamspark pkexec[12979]: bcclsn: Executing command [USER=root] [TTY=unknown] [CWD=/home/bcclsn/.local/share/gnome-shell/extensions/cpupower@mko-sl.de] [COMMAND=/usr/local/bin/cpufreqctl-bcclsn --backend intel_pstate --format json max set 100]
mag 23 15:57:08 dreamspark systemd[12642]: Started Application launched by gnome-shell.
mag 23 15:57:14 dreamspark dbus-daemon[12658]: [session uid=1000 pid=12658] Activating service name='org.gnome.Nautilus' requested by ':1.11' (uid=1000 pid=12699 comm="/usr/bin/gnome-shell ")
mag 23 15:57:16 dreamspark gnome-shell[12699]: meta_window_set_stack_position_no_sync: assertion 'window->stack_position >= 0' failed`

**Comment by fin-ger** [comment]
@bcclsn ah perfect 👍 I'll get to this asap, looks like I messed up the profile management.

**Comment by fin-ger** [comment]
@bcclsn I reworked the initialization procedure. The issue arises when Auto-Switch is enabled and the profile is switched to the configured one before the menu was completely created (due to async stuff in the code). It's quite hard to test for me in VM's as the AC events cannot easily be triggered (or at least I don't know how). Can you test the latest commit of the `feature/modular-cpufreqctl-backends` branch and see if the issue is resolved?

**Comment by bcclsn** [comment]
@fin-ger hi thanks for all! 
it seems to work but not very well ... every now and then it does not switch the profile automatically, I don't know if it is slow to do it or if it is blocked. it doesn't seem to change profile automatically when I manually set one. what test can i do to help you understand?

**Comment by fin-ger** [comment]
Hmm, I'll see if I can get this reproduced on real hardware. I get back to you as soon as I have some progress on this.

**Comment by fin-ger** [comment]
@bcclsn I could reproduce this issue (I think) on my X260 laptop.

The Auto-Switch works fine when already logged in. Unplugging and plugging the AC-adapter in an active Gnome session will switch the profiles reliably.

However, when logging in (the initial setup after successful login) the profile is not reliably switched. The profile is only applied partially (e.g. only the turbo boost or minimum frequency setting is applied). Is this the behavior you observed?

**Comment by fin-ger** [comment]
@bcclsn I think I fixed this with the latest commit (7d93fe9).

@martin31821 @mastercaution The issues are all caused by the callback-based async-io in glib which is a pain to use in JS. I have not investigated if we can safely switch to a Promise based async-await. I fear that old Gnome platforms do not support this. As an example, this is how I fixed the aforementioned race condition:

```diff
     applyProfile(profile) {
         this.minVal = profile.MinimumFrequency;
-        this.updateMin();
-
-        this.maxVal = profile.MaximumFrequency;
-        this.updateMax();
-
-        this.isTurboBoostActive = profile.TurboBoost;
-        this.updateTurbo();
-
-        this.updateUi();
+        this.updateMin(() => {
+            this.maxVal = profile.MaximumFrequency;
+            this.updateMax(() => {
+                this.isTurboBoostActive = profile.TurboBoost;
+                this.updateTurbo(() => {
+                    this.updateUi();
+                });
+            });
+        });
     }
```

This is really ugly and with more lines of code, the nesting gets worse and worse.

Either way, porting the extension to proper async-await is out-of-scope for PR #151 and should be implemented separately (preferably after #151 is merged :laughing:) and definitely not a requirement for Gnome 40 support which is urgent for now. However, for the sake of maintainability, we should discuss and investigate on async-await support in the different Gnome platforms after Gnome 40 support is finished.

**Comment by bcclsn** [comment]
@fin-ger yes, that was the problem and it seems solved with the latest commit. good job!!!
I will test in the next few days :)

**Comment by schaze** [comment]
If it helps for test coverage, I am running an Arch Linux installation with Gnome 40.2.0. I can compile the version in the feature branch _feature/modular-cpufreqctl-backends_, install it and it works just fine. I have a Dell Laptop with a Intel® Core™ i7-9750H CPU.

**Comment by fin-ger** [comment]
@schaze ah perfect :+1: Can you confirm that the frequency setting resets when opening the preferences window? I think this bug should also appear in Gnome 40.

Otherwise, I need someone to review PR #151, although not all issues are fixed by now (see the PR for details). I especially need people to try to break the extensions security model and see if e.g. privilege escalation is possible. I reworked a lot of the backend and do not want to throw anything in production without at least a basic level of security inspection.

Thanks in advance for any help!

**Comment by eleius** [comment]
I am also running Arch Linux with Gnome 40.2, but when I make & install the version in feature/modular-cpufreqctl-backends, there seems to be a visual problem with the frequency sliders. See [screenshots here](https://imgur.com/a/EEQy9Za) (first image: just opened the menu, second image: trying to reduce maximum frequency by dragging the slider, but it also appears on the min freq line)

**Comment by fin-ger** [comment]
@eleius This is expected as the minimum frequency cannot be above the maximum frequency. The bars indicate the minimum and maximum boundaries.

Any idea on how to make this visually more clear?

**Comment by eleius** [comment]
@fin-ger ah, sorry! I thought it was a visual glitch because the max freq boundary was missing but it was shown above (min freq). Maybe just add the boundary to both, as for the leftmost boundaries?

**Comment by fin-ger** [comment]
@eleius Technically, the maximum frequency slider has no maximum boundary. The maximum frequency slider defines the maximum boundary of the minimim slider. The maximum frequency itself can be anything greater than the minimum frequency, therefore there is no maximum boundary.

But I do get your confusion. I just don't know how to visualize this better.

**Comment by eleius** [comment]
@fin-ger Perhaps using a different color (e.g. light gray) for the other slider's boundary would make it clearer that it's just a reference and not the "active" marker? (I don't know if it's possible to use different colors)

**Comment by fin-ger** [comment]
@eleius Yes, it is. We initially used a grey-tone, but it was close to invisible using the high-contrast theme of GNOME. We chose the blue one as it had best visibility using different themes. More on that topic in #142

**Comment by schaze** [comment]
@fin-ger I am not sure if I am testing it wrong, but my frequency settings do not reset.

Testcase:
1. Open menu, change Frequency settings (either using a profile or the sliders directly)
2. Close Menu, reopen menu and open preferences or directly open preferences in opened menu after frequency change
3. Close Preference window
4. Open menu, check frequency settings --> still the same


Not sure about PR #151 and security, this is not really my level of expertise. I will try and run the test series as described in the PR an report back.

**Comment by ndm767** [comment]
I was having an issue on Ubuntu 21.04 with GNOME 40 on X11 where when I installed the extension, it would say Error: UNKNOWN in the drop-down menu. What seems to be causing this is that there are 2 installations of `cpufreqctl` installed, one at `/usr/bin/cpufreqctl` and one at `/usr/local/bin/cpufreqctl-[user]` where user is my username. The second one doesn't seem to support the `--backend` and `--format` flags. The way I fixed this was by doing `sudo mv /usr/local/bin/cpufreqctl-[user] /usr/local/bin/cpufreqctl-[user]-old` and then doing `ln -s /usr/bin/cpufreqctl /usr/local/bin/cpufreqctl-[user]` and then logging out and in after which it seems to work.

**Comment by fin-ger** [comment]
@ndm767 Please see #156 for explanation on why this is not an issue.

**Comment by sharkguto** [comment]
i am getting the following error too

The settings of extension cpupower@mko-sl.de had an error:
```
Gtk.BuilderError: /home/gustavo/.local/share/gnome-shell/extensions/cpupower@mko-sl.de/data/cpupower-preferences.glade:88:1 Invalid object type 'GtkButtonBox'
```

Stack trace:
```
CPUPowerPreferences@/home/gustavo/.local/share/gnome-shell/extensions/cpupower@mko-sl.de/src/preferences.js:53:22
buildPrefsWidget@/home/gustavo/.local/share/gnome-shell/extensions/cpupower@mko-sl.de/prefs.js:41:23
_init@resource:///org/gnome/Shell/Extensions/js/extensionsService.js:209:40
OpenExtensionPrefsAsync/<@resource:///org/gnome/Shell/Extensions/js/extensionsService.js:122:28
asyncCallback@resource:///org/gnome/gjs/modules/core/overrides/Gio.js:115:22
run@resource:///org/gnome/Shell/Extensions/js/dbusService.js:177:20
main@resource:///org/gnome/Shell/Extensions/js/main.js:19:13
run@resource:///org/gnome/gjs/modules/script/package.js:206:19
start@resource:///org/gnome/gjs/modules/script/package.js:190:8
@/usr/share/gnome-shell/org.gnome.Shell.Extensions:1:17
```

**Comment by sharkguto** [comment]
> i am getting the following error too
> 
> The settings of extension [cpupower@mko-sl.de](mailto:cpupower@mko-sl.de) had an error:
> 
> ```
> Gtk.BuilderError: /home/gustavo/.local/share/gnome-shell/extensions/cpupower@mko-sl.de/data/cpupower-preferences.glade:88:1 Invalid object type 'GtkButtonBox'
> ```
> 
> Stack trace:
> 
> ```
> CPUPowerPreferences@/home/gustavo/.local/share/gnome-shell/extensions/cpupower@mko-sl.de/src/preferences.js:53:22
> buildPrefsWidget@/home/gustavo/.local/share/gnome-shell/extensions/cpupower@mko-sl.de/prefs.js:41:23
> _init@resource:///org/gnome/Shell/Extensions/js/extensionsService.js:209:40
> OpenExtensionPrefsAsync/<@resource:///org/gnome/Shell/Extensions/js/extensionsService.js:122:28
> asyncCallback@resource:///org/gnome/gjs/modules/core/overrides/Gio.js:115:22
> run@resource:///org/gnome/Shell/Extensions/js/dbusService.js:177:20
> main@resource:///org/gnome/Shell/Extensions/js/main.js:19:13
> run@resource:///org/gnome/gjs/modules/script/package.js:206:19
> start@resource:///org/gnome/gjs/modules/script/package.js:190:8
> @/usr/share/gnome-shell/org.gnome.Shell.Extensions:1:17
> ```

I am using gnome 40 in manjaro linux

**Comment by fin-ger** [comment]
@sharkguto Please read the thread before posting to prevent unnecessary noise: https://github.com/martin31821/cpupower/issues/145#issuecomment-844933005

**Comment by evgnor86** [comment]
Settings don't work on Gnome 40

**Comment by fin-ger** [comment]
Will be fixed in next update

**Comment by Neo-29** [comment]
What about gnome 41? broken too

**Comment by fin-ger** [comment]
I don't know, I haven't tested it, yet. Have you tested the current master branch with a Gnome 41?

**Comment by Neo-29** [comment]
> I don't know, I haven't tested it, yet. Have you tested the current master branch with a Gnome 41?

in fact it works. I rebuilt from AUR

Unfortunately, I have a leak on CPU usage about 1% . sure it's the extension, I tested witout.

May I open an issue?

**Comment by Neo-29** [comment]
> Can you disable `Show current frequency` in the settings and check if the CPU usage is still increased?

Yep...that solves the issue

But I like to see the frequency

**Comment by fin-ger** [comment]
Yes, but we need the CPU to calculate the frequency and display it in the top bar :sweat_smile: Although, we increased the complexity for how the frequency is fetched. This has something to do with the new abstraction layer supporting AMD Ryzen and Intel Core CPUs. We might improve on that in the future, but for now I would like to leave it as is, to not further delay the Gnome 40 release.

**Comment by Neo-29** [comment]
> Yes, but we need the CPU to calculate the frequency and display it in the top bar 😅 Although, we increased the complexity for how the frequency is fetched. This has something to do with the new abstraction layer supporting AMD Ryzen and Intel Core CPUs. We might improve on that in the future, but for now I would like to leave it as is, to not further delay the Gnome 40 release.

I get it... No pb.
It was not the case before... I just noticed that, I just wanted to let you know the issue 😊

**Comment by fin-ger** [comment]
I really appreciate it! It's really hard to test for all the different systems out there!

May I ask how many CPU cores your CPU has? This might have something to do with the increase.

Also, I'll create a tracking issue for this, so we can track progress on performance improvements for frequency fetching.

**Comment by Neo-29** [comment]
I have 4 cpu on a N4100

**Comment by developomp** [comment]
Is there an ETA for when the extension will be fully gnome-40 compatible?

**Comment by fin-ger** [comment]
We are currently waiting for translations of the new release (#160). If you want you can help and speed things up.

**Comment by fin-ger** [comment]
This extension is now available for dozens of linux distributions (including Gnome 40 support). Please consult the readme on how to install it with your package manager.
